src/hotspot/os/linux/os_linux.cpp

```diff
@@ -3918,139 +3918,98 @@ static void warn_on_large_pages_failure(char* req_addr, size_t bytes,
-  size_t large_page_size = os::large_page_size();
-  assert(bytes >= large_page_size, "Shouldn't allocate large pages for small sizes");
-
+char* os::Linux::reserve_memory_special_huge_tlbfs(size_t bytes,
+                                                   size_t alignment,
+                                                   char* req_addr,
+                                                   bool exec) {
+  assert(UseLargePages && UseHugeTLBFS, "only for Huge TLBFS large pages");
   assert(is_aligned(req_addr, alignment), "Must be");
-  assert(is_aligned(bytes, alignment), "Must be");
-
-  // First reserve - but not commit - the address range in small pages.
-  char* const start = anon_mmap_aligned(req_addr, bytes, alignment);
-
-  if (start == NULL) {
-    return NULL;
-  }
-
-  assert(is_aligned(start, alignment), "Must be");
-
-  char* end = start + bytes;
-
-  // Find the regions of the allocated chunk that can be promoted to large pages.
-  char* lp_start = align_up(start, large_page_size);
-  char* lp_end   = align_down(end, large_page_size);
-
-  size_t lp_bytes = lp_end - lp_start;
-
-  assert(is_aligned(lp_bytes, large_page_size), "Must be");
+  assert(is_aligned(req_addr, os::large_page_size()), "Must be");
+  assert(is_aligned(alignment, os::vm_allocation_granularity()), "Must be");
+  assert(is_power_of_2(os::large_page_size()), "Must be");
+  assert(bytes >= os::large_page_size(), "Shouldn't allocate large pages for small sizes");
 
-  if (lp_bytes == 0) {
-    // The mapped region doesn't even span the start and the end of a large page.
-    // Fall back to allocate a non-special area.
-    ::munmap(start, end - start);
+  // We only end up here when at least 1 large page can be used.
+  // If the size is not a multiple of the large page size, we
+  // will mix the type of pages used, but in a decending order.
+  // Start of by reserving a range of the given size that is
+  // properly aligned. If a requested address is given it will
+  // be used and it must be aligned to both the large page size
+  // and the given alignment. The larger of the two will be used.
+  size_t required_alignment = MAX(os::large_page_size(), alignment);
+  char* const aligned_start = anon_mmap_aligned(req_addr, bytes, required_alignment);
+  if (aligned_start == NULL) {
     return NULL;
   }
 
-  int prot = exec ? PROT_READ|PROT_WRITE|PROT_EXEC : PROT_READ|PROT_WRITE;
-  int flags = MAP_PRIVATE|MAP_ANONYMOUS|MAP_FIXED;
-  void* result;
+  // Start of by mapping large pages.
+  size_t large_bytes = align_down(bytes, os::large_page_size());
+  char* large_mapping = reserve_and_commit_special(large_bytes, os::large_page_size(), aligned_start, exec);
 
-  // Commit small-paged leading area.
-  if (start != lp_start) {
-    result = ::mmap(start, lp_start - start, prot, flags, -1, 0);
-    if (result == MAP_FAILED) {
-      ::munmap(lp_start, end - lp_start);
-      return NULL;
-    }
+  if (bytes == large_bytes) {
```

Comment:
Shouldn't we do the check `if (large_mapping == NULL) {` before this?

Reply:
We could, and I thought about adding a comment here. I probably should have. The reason we don't do the explicit check it is that if `large_mapping == NULL` then NULL will be returned as expected and there is no additional work to be done if `bytes == large_bytes`.

Reply:
When `bytes == large_bytes` don't we need to `unmap` the original reservation in case `reserve_and_commit_special`  fails i.e `large_mapping == NULL`? 

Below we have the comment 
// Large mapping failed, so we need to unmap the reminder
// of the orinal reservation.

Maybe I am missing something.

Reply:
ok, now I get it. I had missed that when mmap fails, the original reservation is unmapped for this range so no need to do any additional clean up.